### ISSUE TITLE
[OpenApi] lossy generation of java client from tapir generated yaml

### ISSUE THREAD

**Issue by ghostbuster91** [softwaremill/tapir#308]
Reported by @runebarikmo on gitter - https://gitter.im/softwaremill/tapir?at=5dce9744c26e8923c42c45b8

TLDR: 
Generating a java client through http://editor.swagger.io/ from tapir generated yaml, where tapir model contained coproducts, results in model with loosely coupled classes without a common ancestor. 

The question is whether we should just add `allOf` to our model or it is some bug in an official client generator.

**Comment by adamw** [comment]
Should be fixed in 0.12.4

**Comment by runebarikmo** [comment]
I've upgraded to 0.12.4 and it looks much better now. Thanks guys :)

**Comment by adamw** [comment]
As @gbastkowski pointed out in https://github.com/softwaremill/tapir/issues/354, using both `allOf` and `oneOf` produces weird schemas.

And this kind of makes sense. If we have a schema like:

```
   Animal:
      oneOf:
      - $ref: '#/components/schemas/Boar'
      - $ref: '#/components/schemas/Cat'
    Boar:
      allOf:
      - $ref: '#/components/schemas/Animal'
      - required:
        - weight
        type: object
        properties:
          weight:
            type: integer
    Cat:
      allOf:
      - $ref: '#/components/schemas/Animal'
      - required:
        - name
        - weight
        type: object
        properties:
          name:
            type: string
          weight:
            type: integer
```

this says: `Animal` is one of `Cat` or `Boar`. Cat is some properties, AND all of `Animal`. Which means that `Cat` is some properties, AND one of `Cat` or `Boar`. Which doesn't make sense: a `Cat` is a `Cat`, not a `Cat AND Cat or Boar`

Looking at https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/, it looks like the correct way to implement inheritance is by using `oneOf` **only**. In this document, `allOf` is mentioned only as a way to group common properties.

So this would seem that it's a bug in the client generator, not in tapir? 

I would thus revert #324 and bring back the previous behavior.

**Comment by gbastkowski** [comment]
Here's the project that demonstrates that behaviour: https://github.com/gbastkowski/tapir-demo

**Comment by gbastkowski** [comment]
Having read the specification and also had a look at the SwaggerUI rendering,
I come to the conclusion that `allOf` mostly exists to reduce lines of YAML code.
A strong selling point if the YAML file is crafted by hand.

When talking about YAMLs that are generated from Scala case classes, it seems to be a good decision for most use cases not to support it.

**Comment by adamw** [comment]
Released in 0.12.19 - please check! :)

**Comment by gbastkowski** [comment]
Worx. Thanks a lot.
